src/functions/env.ts

```diff
@@ -0,0 +1,238 @@
+    );
+  }
+  if (key.startsWith("X_GOOGLE_") || key.startsWith("FIREBASE_")) {
+    throw new Error("Key starts with a reserved prefix (X_GOOGLE_ or FIREBASE_)");
+  }
+}
+
+// Parse dotenv file, but throw errors if:
+//   1. Input has any invalid lines.
+//   2. Any env key is reserved.
+function parseStrict(data: string): Record<string, string> {
+  const { envs, errors } = parse(data);
+
+  if (errors.length) {
+    logger.debug(`Invalid dotenv file, error on lines:\n${errors.join("\n")}`);
+    throw new Error(`Invalid dotenv file, error on lines: ${errors.join(",")}`);
+  }
+
+  const invalidKeys: Record<string, string> = {};
+  for (const key of Object.keys(envs)) {
+    try {
+      validateKey(key);
+    } catch (err) {
+      invalidKeys[key] = err.message || "unknown error";
+    }
+  }
+  if (Object.keys(invalidKeys).length > 0) {
+    logger.debug(
+      `Found invalid keys: ${Object.entries(invalidKeys)
+        .map(([k, v]) => `${k}: ${v}`)
+        .join("\n")}`
+    );
+    throw new Error(`Found invalid keys: ${Object.keys(invalidKeys).join(",")}`);
+  }
+
+  return envs;
+}
+
+/**
+ * Loads environment variables for a project.
+ *
+ * Load looks for .env files at the root of functions source directory
+ * and loads the contents of the .env files.
+ *
+ * .env files are searched and merged in the following order:
+ *
+ *   1. .env
+ *   2. .env.<project or alias> (or .env.local if isEmulator is true).
+ *
+ * If both .env.<project> and .env.<alias> files are found, an error is thrown.
+ *
+ * @return {Record<string, string>} Environment variables for the project.
+ */
+export function load(options: {
+  functionsSource: string;
+  projectId: string;
+  projectAlias?: string;
+  isEmulator?: boolean;
+}): Record<string, string> {
+  const targetFiles = [".env", `.env.${options.projectId}`];
+  if (options.isEmulator) {
+    targetFiles.push(".env.local");
+  } else {
+    targetFiles.push(`.env.${options.projectId}`);
+    if (options.projectAlias && options.projectAlias.length) {
+      targetFiles.push(`.env.${options.projectAlias}`);
+    }
+  }
+
+  const targetPaths = targetFiles
+    .map((f) => path.join(options.functionsSource, f))
+    .filter(fs.existsSync);
+
+  // Check if both .env.<project> and .env.<alias> exists.
+  if (targetPaths.some((p) => path.basename(p) === `.env.${options.projectId}`)) {
+    if (options.projectAlias && options.projectAlias.length) {
+      for (const p of targetPaths) {
+        if (path.basename(p) === `.env.${options.projectAlias}`) {
+          throw new FirebaseError("Can't have both .env.<project> and .env.<alias> files.");
+        }
+      }
+    }
+  }
+
+  let envs: Record<string, string> = {};
+  for (const targetPath of targetPaths) {
+    try {
+      const data = fs.readFileSync(targetPath, "utf8");
+      envs = { ...envs, ...parseStrict(data) };
+    } catch (err) {
+      logger.debug(`Failed to load environment variables from ${targetPath}`, err);
```

Comment:
The error will get printed anyway when bubbled up, so why a debug log line?

Reply:
Good point. Originally, debug line contained more info, but I decided to push it all on to the error as well.